### PR TITLE
A Flux model

### PR DESCRIPTION
**Pull request by rossviljoen**
Adds a high level API for a model which can easily be used with Flux (although doesn't add a dependency on Flux itself).

Examples have been updated to use the model.

It should also be straightforward to adapt the model to support something like [ParameterHandling.jl](https://github.com/invenia/ParameterHandling.jl) for use with general optimisers - this will probably be left to a future PR though.

### CODE REVIEW

File: src/models.jl [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    z::AbstractVector           # inducing points
+end
+
+@functor SVGPModel (k, m, A, z,)
+
+function SVGPModel(
+    kernel_func,
+    kernel_params,
+    inducing_inputs;
+    q_μ::Union{AbstractVector, Nothing}=nothing,
```
Comment (theogf) [inline src/models.jl:21]: As a general formatting rule, there should be no spaces in `Union`

File: src/utils.jl [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# These methods to create a Cholesky directly from the factorisation will be in Julia 1.7
+# https://github.com/JuliaLang/julia/pull/39352
+LinearAlgebra.Cholesky(L::LowerTriangular{T}) where {T} = Cholesky{T,typeof(L.data)}(L.data, 'L', 0)
+LinearAlgebra.Cholesky(U::UpperTriangular{T}) where {T} = Cholesky{T,typeof(U.data)}(U.data, 'U', 0)
```
Comment (theogf) [inline src/utils.jl:4]:
```suggestion
if VERSION < v"1.7"
    LinearAlgebra.Cholesky(L::LowerTriangular{T}) where {T} = Cholesky{T,typeof(L.data)}(L.data, 'L', 0)
    LinearAlgebra.Cholesky(U::UpperTriangular{T}) where {T} = Cholesky{T,typeof(U.data)}(U.data, 'U', 0)
end
```

File: src/models.jl [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    return GP(kernel)
+end
+
+function prior(m::SVGPModel)
+    kernel = m.kernel_func(m.k)
+    return LatentGP(GP(kernel), m.lik, m.jitter)
+end
+
+function loss(m::SVGPModel, x, y; n_data=length(y))
+    return -elbo(m, x, y; n_data)
```
Comment (theogf) [inline src/models.jl:80]: we could define `vfe = - elbo` :D (VFE is the Variational Free Energy which is exactly the negative elbo)

File: src/models.jl [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    return SparseGPs.elbo(fx, y, fu, q; n_data)
+end
+
+function _init_variational_params(n)
+    m = zeros(n)
+    A = Matrix{Float64}(I, n, n)
+    return m, A
+end
+
+function _init_variational_params(
```
Comment (theogf) [inline src/models.jl:94]: Should we add a keyword argument to enforce the eltype and type of \mu and \Sigma?

**Review by theogf — COMMENTED** [review]
Did not review everything in details but here are some comments.
Additionally, should we have some internal `train` and `predict` functions? We could add a dependency on Flux via `Requires.jl`?

File: src/models.jl (hunk 3 above)
Comment (rossviljoen) [reply src/models.jl:80]: do you mean rename `loss` to `vfe`?

File: src/models.jl (hunk 4 above)
Comment (rossviljoen) [reply src/models.jl:89]: I added an eltype parameter, but I can't think of a reason why you wouldn't want to use `Vector` / `Matrix`, so maybe a type parameter isn't needed?

File: src/models.jl (hunk 3 above)
Comment (rossviljoen) [reply src/models.jl:81]: do you mean rename `loss` to `vfe`?

**Comment by rossviljoen** [comment]
> Did not review everything in details but here are some comments.
> Additionally, should we have some internal `train` and `predict` functions? We could add a dependency on Flux via `Requires.jl`?

Training works with Flux's default training loop, so I'm not sure an internal `train` function would even make things much easier (besides setting defaults perhaps).

File: src/utils.jl [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# These methods to create a Cholesky directly from the factorisation will be in Julia 1.7
```
Comment (st--) [inline src/utils.jl:1]: :+1:

File: src/models.jl (hunk 4 above)
Comment (theogf) [reply src/models.jl:89]: Well in the crazy case where someone would like to use GPUs, one would need pass CuArrays but I guess they would just have to make the effort of passing their own variational parameters.

File: src/models.jl (hunk 3 above)
Comment (theogf) [reply src/models.jl:81]: I meant more simply to have an alias:
```
vfe(args...; kwargs...) = -elbo(args...; kwargs...)
```
but maybe @st-- and @willtebbutt think it's not necessary

File: src/models.jl (hunk 4 above)
Comment (rossviljoen) [reply src/models.jl:89]: Since it's a functor, isn't it easiest to just use `Flux.gpu` for that after it's initialised?

**Comment by rossviljoen** [comment]
Pretty much everything useful in this PR is in the regression/classification examples already, so closing

**Pull request closed without merging**